### ISSUE TITLE
Type narrowing not applied in nested function with assert

### ISSUE THREAD

**Issue by parched** [microsoft/pyright#4802]
While upgrading pyright I noticed this new error which I believe should not be an error.

**Screenshots or Code**
![image](https://user-images.githubusercontent.com/5975405/226153166-2af0517f-664b-4146-9a63-4b0b8b53207f.png)

```python
def _get_iterable() -> list[str] | tuple[str, ...] | None:
    return []


def a_bug() -> str | None:
    result = _get_iterable()

    if result is None:
        return None

    def get_first() -> str | None:
        assert isinstance(result[0], str)
        return result[0]

    return get_first()

```

Related https://github.com/microsoft/pyright/issues/1731

**VS Code extension or command-line**
Commandline, started happening in 1.1.274

**Comment by erictraut** [comment]
This is by design. In general, variables that are captured by an inner scope do not retain any narrowing performed by an outer scope. Pyright allows the narrowing to be retained in limited cases where it can be proven to be safe. By contrast, mypy never retains the narrowing. In your case, you're using the variable in an index expression, and it's difficult (and generally impossible) for a static type checker to determine whether the narrowing is safe to retain. You can assign the narrowed value to a temporary variable if you want this to work without a type error.

```python
def a_bug() -> str | None:
    result = _get_iterable()

    if result is None:
        return None

    x = result

    def get_first() -> str | None:
        assert isinstance(x[0], str)
        return x[0]

    return get_first()
```
